src/ApplicationRecordLog.cpp

```diff
@@ -169,21 +158,90 @@ namespace geopm
 
     void ApplicationRecordLogImp::epoch(const geopm_time_s &time)
     {
-        check_setup();
         std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
         m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
         check_reset(layout);
 
         ++m_epoch_count;
         record_s epoch_record = {
-           .time = geopm_time_diff(&m_time_zero, &time),
+           .time = time,
            .process = m_process,
            .event = EVENT_EPOCH_COUNT,
            .signal = m_epoch_count,
         };
         append_record(layout, epoch_record);
     }
 
+    void ApplicationRecordLogImp::cpuset_changed(const geopm_time_s &time)
+    {
+        auto cpu_set = m_scheduler->proc_cpuset(m_process);
+        int num_cpu = m_scheduler->num_cpu();
+        for (int cpu_idx = 0; cpu_idx < num_cpu; ++cpu_idx) {
+            if (CPU_ISSET(cpu_idx, cpu_set.get())) {
+                affinity(time, cpu_idx);
+            }
+        }
+    }
+
+    void ApplicationRecordLogImp::affinity(const geopm_time_s &time, int cpu_idx)
+    {
+        std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
+        m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
```

Comment:
I think we could consider a refactor here to use dynamic_cast instead of explicit C style casting.